[clang] Fix ASTWriter crash after merging named enums

michael-jabbour-sonarsource opened the pull request:
Clang already removes parsed enumerators when merging typedefs to anonymous enums. This is why the following example decl used to be handled correctly while merging, and ASTWriter behaves as expected:

```c
typedef enum { Val } AnonEnum;
```
However, the mentioned mechanism didn't handle named enums. This leads to stale declarations in `IdResolver`, causing an assertion violation in ASTWriter ``Assertion `DeclIDs.contains(D) && "Declaration not emitted!"' failed`` when a module is being serialized with the following example merged enums:

```c
typedef enum Enum1 { Val_A } Enum1;
enum Enum2 { Val_B };
```

The PR applies the same mechanism in the named enums case.

Additionally, I dropped the call to `getLexicalDeclContext()->removeDecl` as it was causing a wrong odr-violation diagnostic with anonymous enums.

Might be easier to to review commit by commit. Any feedback is appreciated.

### Context

This fixes frontend crashes that were encountered when certain Objective-C modules are included on Xcode 16. For example, by running `CC=/path/to/clang-19 xcodebuild clean build` on a project that contains the following Objective-C file:

```c
#include <os/atomic.h>

int main() {
  return memory_order_relaxed;
}
```
This crashes the parser in release, when ASTReader tries to load the enumerator declaration.

michael-jabbour-sonarsource commented:
Ping?

Note that this PR fixes a crash in ASTWriter on valid Objective-C modules code; The initial PR title didn't accurately reflect that.

ChuanqiXu9 commented:
@jansvoboda11 for OC related things

jansvoboda11 commented:
> @jansvoboda11 for OC related things

Thanks! CC @vsapsai and @ian-twilightcoder.

michael-jabbour-sonarsource commented:
Gentle ping :smile:

The aim of the PR is to fix an ASTWriter crash that currently happens when serializing merged enums. Such a case was encountered while parsing Objective-C code that uses the XCode 16 SDK. If any of the changes look risky, I am happy to rework them and/or add more tests to increase confidence.

Thanks in advance.

michael-jabbour-sonarsource commented:
Gentle ping :smile:

I would love to hear your feedback, and if any changes are needed, I am happy to address them promptly.

vsapsai commented:
Sorry for the delay. Will need to look at the change in a debugger.

michael-jabbour-sonarsource commented:
Thanks for the update @vsapsai. Sounds good. Let me know if there are any changes that can make reviewing the PR easier!

michael-jabbour-sonarsource commented:
Gentle ping, and I wish you all a happy new year. :smile:

cor3ntin commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -2551,18 +2551,7 @@ void Sema::MergeTypedefNameDecl(Scope *S, TypedefNameDecl *New,
-          assert(EnumScope->isDeclScope(ED));
-          EnumScope->RemoveDecl(ED);
-          IdResolver.RemoveDecl(ED);
-          ED->getLexicalDeclContext()->removeDecl(ED);
```
I am not super confident about this change, is it related or could it be separate? @AaronBallman

michael-jabbour-sonarsource replied:
I can try to dig deeper to answer the points I mentioned (especially point 2) if desired, but I would need to allocate some time (maybe at the end of the week).

Any pointers/hints/thoughts would be appreciated.

michael-jabbour-sonarsource replied:
I had a quick look into this today, and I could see that some other ODR checks aren't emitted unless `-ast-dump-all` is provided. For example, in [this CE example](https://godbolt.org/#z:OYLghAFBqd5TKALEAXATgU0wGgJYgQDGAtgIYDWmAgusAM4gDkAtACIDCAstQNICiAfQBCAVQCSAGTaCAKgE0ACvwCkAZjZtMAIwCuwJjiIB7EgAc8AG0zpJZAHbBdZYJnEATEERUAmYb%2BEjXXpUUwB5XVQzSIAxK0x7MhJMNEwQnAAzeMZYTHc8UPQPEB9M%2BOK1MutE5JB6JDIsdwA6JBxQJkM8enF7IktddzzmAAYu%2Bi4yPHsAZWNddCIUpgBGcbCzBNHDSwdgYs6AShwIPILjIs9KrOtigBYqhKSUzAAPJLNrZqIzM3bmca9fqDYZMMb4CZTWbzRbLcHdDZbME7PYHJjHU75QrFFaPYoANkeNRSJGM7l0X1J7nIfw6gL6AyGnmREMm0zmCyWzDWEMR9m2OF2jmK3j8AQxZ2xngeNzcngA7ETniAuGTqHFrCtWv9OhCgYzQfDIeyYVzVutNvyWUL9p5Rf4/BKsRdigBWPGeAAcStqqvc6viPm1dL1DJBzKNbOhnOWPIRloFNpFAXFJ0lLs8hNlxQAnD6Un7hBrMMGAaHgUyBd0oxzYdyLUjwUm7SnHcdutxKJhFOhjAArTBEVD1nB9kC4%2ByYADuxZxI2OlmYrsMVvBxmYsiwmAA1L41Dzh8iMSgTEN%2BM70PhCLLiSOMsZ7KgZkQyNZxzKH6hRPRu68BWP3RtLxW0CMwHEwAA5ZU/QDawgzaH9rCHPAH0IBJ3A4YwBhIflcXQyRpkg3QSG0GwShwMxjHoAoUPsTDsNwiiqJoh8CMnCDiNI9ByMQwdUFomZUEaVB6OIxjeOQh9BOEtiiJIsjShCYTRJw8ccCU9BUFkjj5O4nxjnoU1lgAemM7d%2BHMVAAE9t1lFQRmodFBSXFcBXXGDi3gnBDzGY8vDJTBz3OS8CAgG9lUMmNWl%2BTJP2fV8UhWD9H2/X9/3IoD7QCCjwKg2p6kaPJtQk/jUNOewMKwsS1PwwidK48jKOo0q6Kq1TcSalj7G0ziFPUzAkJa6TNJU8SBr4gShM0nrdJ4qaRLa8T5pmhr9PUozmASYjty4Kz%2BHsYifF3eVhB2qyADVXxEY62HUfwHKcxcmGXHBV0MdcCqabUfKciATwCoLCivML4lvchpm%2BGL70feK3yS2KUp/RQ/xZMdKkykCcvY5U3g%2BL4fj%2BEraLQirRpKUpavY3ruMqTqWrJnxSjp2iVrIyoiak%2BaGcU8bJNmZa6upkB2a5xbyfUgWqdmtQDI2pg92mCsd18HxC2LVoVfs6gtemVBt3B%2BwIEOY7TqwVAFnsbcODum6tcelzXrckBcfMfGYp%2BvzT0Ci9gfC2pzQRp8Xzh5KvyRlHwTHXEMbFPwscg5UqQpEsqRpfrBuJ8rKoYmqKtZ7iVkJZmHzJjrmJa/Px0JDn%2BeUsXcRr4atMF2bC4luuc4byW5IawvZZjZgk%2BsHa1RNrXt31slk5H/1ixWMeHIniekEwMghnQXdGY8%2BItSQTXF6Xt5KM07cACpx9tg%2Bh53be4IX6gl%2B3Fe15sTfVbVTyNcZi%2BJ6Pi49fPgfFQ8pboOWAaA7WDlr4z1OsA%2B6D8n6r3Xm/NW8Qv4%2BAvn/E%2BgDtYgLtguB2b0cDrmvs0VOZA/gexOP9M8PtQp%2B1jIYaGQcErvkDqlZG6UHgxwdKBXK0EP472KrzFqJNs7VTwnnFuDVabl1oqXJizUWbSLZunCanMO4SLUXzJulcRaaPau3aaKiab9zrPLBy1lNhDAyNuLaJATZWxuuZO6%2BDnLPVciydygjNTfW2IcY4z916EGOBQEAroRjuLuIYEg4TIlEPXLqRg7oqFwFgDQ72wVfag2VAHZhsNEqhw4RHQwY5CQ8OymBbGvoyRFjQQhERmd0IKMpj3MiRc5El3roorqldq6NI0SNbpjdu71XaUYhancJl9LMWaUyMDizbiGJadw9BtwPhnrBTA8F1n2EsFZZoWsFZhiGCgnx2z0FHLAZYqy1jMC2PsY4jgOBtxsGcaoNQ8D7YeMdl4lUtT1ZtA9oEpBZEjZRJiXE35a4yyMEicCnAYSInuLUJ4mFupUmIDScgfy5h4ghUICYPF1huL9D2IIMg9AfyaVovQQQGBdD2AoFUVANhGAQG0AKbQ0xGhWRHNyxI6ArJhG0AOIcI4iXJEfGEPZfKWRDD0MADgr5LCMADkMcgjg3zyrwFgZCAA3NIAo3iDkiIw/Aj4BqJjwNodAvLMJDAFBgPAsSA6GvQNoKimAtCauAJYaYHRfKZF2Awc6eBpx8jvHFYOhT2Hh3StHPYwFY6BH9doHI4LjBmBamqlgrxtxEG3CwDI181nFuviwGkxb4jqDYKQ8hZgi0ZHoFZR8ZBXgsAfPsotAANMljgi2UtQCwck5gWAqveu69AeAhgZoMgM/kWcWlSKlr3Hp9NunF26iYtSIyDFjQzoM5uq6yJd33bu0ZQsVjHAwNgccBCfkJOYNuV4np8QsHxHcAtNptwQEpdSnN24GVMuNhATc2BN77kOIYBFQSwUPuiTgWJrotTynxJ6NQOYVg%2BHxFhnMOZ8TyndE%2BpJIB4X%2BJOFilA8woiRGBkSiwJK8rLFfe%2Bz9369jbhWDmZoWpwT5JjWw5hxSE3eS3DOtSU47VmHrA%2Bl6JGIi0b1sYWxrGP1fv7cAX9/6bCAeAxQUDmFiWvz3NemD/iEOQruJ6ZoPh9z4hGJ6eUahHMjBGGoHwxGnakfI0eUJUKnqouhe9MsCKno%2BDRSFjFFmcDuuoqhO4QA), I have the enum `E` defined with different enumerators, and I don't get a diagnostic unless I have `-Xclang -ast-dump-all` on the command line (removing it, somehow removes the diagnostic).

While I still don't have a concrete answer for what I mentioned above, it seems to me that this is a separate bug in the ODR check implementation, and that ideally, we should get the diagnostic without the `-ast-dump-all`.

If my conclusion above is correct, I would argue that dropping this call is necessary for these ODR checks to work (once the bug above is fixed).

I may try to dig deeper later into why the diagnostics are not emitted without `-ast-dump-all`. I am curious about your thoughts here.

AaronBallman replied:
CC @ChuanqiXu9 @Bigcheese @vgvassilev as serialization and modules maintainers.

vgvassilev replied:
Generally, clang retains as much information about the code as written in its AST invariants as possible. We need to be extra careful when removing content from the lexical declaration context, generally. 

However, for C and I believe ObjC, we do not rely on the lexical structure when making lookups but use the IdResolved chains. I do not think we have done a lot of work to make the resolver chains modules friendly. Can we solve this issue by reusing some level of modules-specific declaration shadowing?

michael-jabbour-sonarsource replied:
Thank you all for looking into this :pray:

> Can we solve this issue by reusing some level of modules-specific declaration shadowing?

I am new to modules, and I only learned how this part of the code works while debugging the crash I am sharing here. I am not entirely sure I understand what this means, could you provide some pointers that I can start from (e.g. function names, commits, ...etc)?

I can try to dig into this at the end of this week or next week if needed...

vgvassilev replied:
@ilya-biryukov, this PR is deletes a statement `ED->getLexicalDeclContext()->removeDecl(ED);` which I think is okay. Is there an easy way to check on your infrastructure if we did not miss some obscure code example which proves us wrong?

ilya-biryukov replied:
I just wanted to circle back and report that the tests are running right now, but may not finish before the end of the working day tomorrow.

Sorry for a slight delay. We should definitely know if this breaks anything tomorrow, but maybe even later today.

ilya-biryukov replied:
No issues on our side with this change. So LG!

michael-jabbour-sonarsource replied:
Thank you all very much for checking this. I will be updating the failing test in a few hours (or tomorrow).

cor3ntin reviewed (commented):
This flew under the radar, sorry about that and thanks for your patience

Just a couple questions/comments. Also, can you add an entry in `clang/docs/ReleaseNotes.rst` ? Thanks

vsapsai commented:
In general, personally I quite like the idea of removing decls from the scope. Though I think when I've tried to do so [in a different context], there were some problems. I'll try to find what I was doing and why it wasn't working. It's not a blocker for your change but it can be helpful to understand potential pitfalls better.

michael-jabbour-sonarsource commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -4034,6 +4034,12 @@ class Sema final : public SemaBase {
+  /// another definition visible.
+  /// This method performs any necessary cleanup on the parser state to discard
+  /// child nodes from newly parsed decl we are retiring.
+  void RetireNodesFromMergedDecl(Scope *S, Decl *New);
```
Thanks for the suggestion, done in f55ca2bdf0e0fabdc672f0f85007d091272a8891.

michael-jabbour-sonarsource commented:
@vsapsai Thank you very much for the input. :pray: 

> Can you trigger the crash without -ast-dump-all?

> Can you use in the test %clang_cc1 instead of %clang?

I addressed both of these in f6f0887d22172c9db4f602b7f272630a97a708bd. I added a separate invocation without `-ast-dump-all` that reproduces the crash prior to this PR. I also switched to `%clang_cc1` instead of `%clang`.

For convenience, you can also find the crashing example here on CE: [link](https://godbolt.org/#z:OYLghAFBqd5QCxAFwE4FN0BoCWIIDGAtgIYDW6AgqsAM4gDkAtACIDCAspQNICiA%2BgCEAqgEkAMi34AVAJoAFXgFIAzCxboARgFdgDLAQD2RAA44ANulTiSAO2DaSwdKIAmIAkoBMg74IPatMjGAPLayCbhAGIW6LYkROgo6EFYAGax9LDorjjBqG4gXumxhSollvGJILQIJBiuAHQIWKAM%2Bji0orYE5tquOYwADB20HCQ4tgDKhtqoBEkMAIyjISZxw/rmdsCF7QCUWBA5eYYF7uUZloUALBVxCUnoAB4JJpaNBCYmrYyj3b1%2BoMGCNcGMJtNZvNFqDOmsNiCtjs9gxDsdcvlCkt7oUAGz3KpJIiGVzaD7E1ykH5tf49PoDdyIsHjSYzOYLRgrMHw2ybLDbeyFTw%2BPxok6Y9x3K4udwAdgJjxAHBJlBiliWzV%2B7TBAPpwNh4NZUI5y1W615TIFu3cwt8PjFGLOhQArDj3AAOBXVZWuVWxLyamk6ulAxkGlmQ9mLLlw818q1Cvyio7ip3ufHSwoATi9SR9gjV6EDf2DgIZfM6EbZ0M5ZoRoITNqT9sOnU45HQ8lQhgAVugCMha1geyBsbZ0AB3QtYoaHcyMZ36C2gwyMaQYdAAam8Ki5g8RaKQRgGvEdqFw%2BGlhJqxqLCG%2B6UMtmQUwIJEso6lT%2BQwloneefIjsUVoeM2/gmHY6AAHKKrU9Q5Jqf6WAOOBPvgcSuGwhh9EQvLYhh4iTNB2hEJoVggK6JiGLQeSobYWE4XhWBUTRyB0YR45QSRZGoBRWBIf2bFPlMyD1MgDEkUxAkocJomoMgHHEaR5GukEYkSbho78XJClEVxym8c6hy0LejAAPRmZu0gIFu47PMgm7IBOhibi8byWLQjmoDgwDOKgnnIDZm4EKgJC1JuGTPDkm6TI5CCdJu8gAEpKEMlDIAAnusAxpK5tgkZuHAZbw%2BVEEs26yoIhUZQAau%2B/CUBVLDVSVJFLKovhpXEBVFa1RBeBVVVFXV5hCE1HWpelWU5OguXdUQg3VSN/BsE1LWlSoE1daVi3DfVzVKLKLBbZQqL8guS58qucENJq%2B4jIeHgkugp6nOeeAQFeiqkJMnwPmk36vu%2BSRLF%2Bz6/v%2BgEgOUIG2n4zGQTB1RuaYHxfD80lCbyxy2Jh2GSVpBF6dx5F3CxtFPhpTHk1jin6TxIB3JjdEiep%2BOadizOyWJdMk7xTM6VTWlqfJvMGYzxmmQwO6TGWW7eF4%2BaFs0CuTZNkwOT9tgQPsi0YMgcy2Et9WrX4xujcdKhVYdx1pWd84MIuWDLvoq4o%2B8Rbo1g91nRAR7Pa9%2BQXp9sTXqaj7PkDH6gxHP5/vIAFMiO2Kw2BCOcd9JJkkWFJUvx6DIVj6G40L%2BG42LDNLPiNN0aXzHURTtgV%2BRVf54XLOC%2BzUkF4JHc88T4utyL4ld8LOnN7xVeS1GjAUtnhUqoNk2bpuc%2BWAvvqFuVh2dY1K%2BbjZJADKg25eIrKpbyrZ/L/vLxUfJm4AFQ3zbN9r1uPp%2BpYA07zfK%2BH8fU%2B59N7%2Bivl4P%2Brlnj3wcs/NKK9X5pQQZQSa78N7W0qjfABVggFK1iGAm%2Bd8zgwLVkdSa9sLrOyuiAd%2BjRc4kB%2BD7R6x4XpnmDl9ao4cAaRzfNHMGcdIZJ0ZvyHYoERQ%2BHTtBRUn9L4tC5tjDCdciacT5kUeurFa6j2xDXJ8E9VFyNZvJOu%2Bjx4DwZsUYeRiTHKPFl4aeNZpZeBULLPUQCboIQQKrO2c4KEuywKuaRsQNQtEYUcf2J5WEfXYdGfQXCXw8JBnwiGCcoaulTmI8CiMpEX1AbInuMl5El00VgJRSkzFqMbnXbRTdTHkXMXkrGBiR6MWFvUvuosam8XMZ3ZpnMrGlNqXYk0FlNy8FMJlCKsQyHeMdpdJk/jsnfzupsJhAcImXlDoqThgN4mfljkkxOoIRz4jSXaDJGdvQkgLHg3J7c0I4zxj04p5cOl8SqZUhutMXmqVadzQxRTjH92sQzVS3SCa9MBf0wygzFjDNwevAY5pXCeSfBvL%2B6AAwIE3E%2BcwGVGhq0cc4oEOCFnovwWlYZup%2BiTGAJuNxTRMU2QwMFEggQUibimOgUgjkXKJBoFuQKW5KBTGkJuWwz1PJpG7AtOlzRNwTjyJigVB8cCuAGEbJ8bLJXGFRXixBBKQwDFcXUW6Hjr5ePOjMyhcylSXOVsE5ZhwsG8R1lgMgFEhgWpuPoIgo4hget8aubU9BsQhLgLAGAiAnqmFiO9fARho2WF4r0HY/Awp/nknRWg/A0DaFsGQCoyArD0AgJoPkmhJj1AykOct8RUAZRCJoPsA4hzxsSM%2BEItgcV8gGDoYAbB3zmHoOHAYpB7AfiZLkDAKEABuKQ%2BQvH7OEaJuBnwF3jDgTQoU61YQGHyNAOAfXh1nagTQ1F0AaFHcAcw1LlnpG2HQGqOBJw8iHLEqOCS9nxwOfoZOwj7CiNOfyDdWQXWGBMFjIdTBnhYsbb3WdTACCbiYGkd%2BnlkPvyYFSZDky1A0LoSYJDaRaAZWfCQZ4TBsUZVdse7yAwQPGR%2BQUh5YKnmuF0Voj5GjHlVPY23XuvymksYBe0oFLdtJs0ecPdjhw0CYFHNMp2AbGCbmeO6XETBcQ3GClaTcEA01WAg15XNZBdYQHXJgU%2Bu59j6EYY69AR9yIuodl6rAPqVDukaI42UuJZRLFlGp507ohg3BUK6JTQaQAetDRGv2IBZgRHCMHeNZhE1I0WKp9TmntM7E3EsLMjQNSgjfTsmOsT9lQxDRuFVWkJyhRMLWBTsyVyMDCAlhyhhcoZY01p5N9hdP6YzU%2BAKqBjOmawgm7BO4ljWe9g6z13qQC4g1N5xxXh3ReCWCoG4eXnT4nC/oegUW5tuudB6h2KgmuuxLLZi1XhLt%2BOu3N49NE0I3CAA%3D%3D).

vsapsai reviewed (commented):
I've realized that the test doesn't have to be Objective-C, the failure is reproducible in C. Curiously, we aren't hitting the assertion in C++/Objective-C++ but I don't know why.

I wanted to try defining the enums in different scopes (e.g. in a struct) and see how it works. Consequences for scopes aren't clear, so that's worth taking another look.

A big surprise for me was that the assertion happens in the stack trace
```
 #7 0x0000000105168664 clang::ASTWriter::WriteCUDAPragmas(clang::Sema&)
 #8 0x00000001033ea29c clang::ASTWriter::WriteFPPragmaOptions(clang::FPOptionsOverride const&)
 #9 0x00000001033e4e70 clang::ASTWriter::WriteIdentifierTable(clang::Preprocessor&, clang::IdentifierResolver*, bool)
```

The fact that `WriteCUDAPragmas` is somehow involved while we have no CUDA pragmas is very suspicious.

vsapsai commented:
For the record, the previous work I've abandoned is https://reviews.llvm.org/D114833 Doesn't seem particularly relevant to this change to be honest.

michael-jabbour-sonarsource commented:
@vsapsai Thanks again for the feedback :pray: 

> I've realized that the test doesn't have to be Objective-C, the failure is reproducible in C. Curiously, we aren't hitting the assertion in C++/Objective-C++ but I don't know why.

I dug a bit into this today, and I could find a few differences:

1. ODR checks seem to be implemented on C and ObjC only (see df0ee34bc2523f11b907311670c776cb3de108c1). As far as I understand, in the C++ case, `Sema::ActOnTag` sets [`SkipBody.ShouldSkip` to true](https://github.com/llvm/llvm-project/blob/cd708029e0b2869e80abe31ddb175f7c35361f90/clang/lib/Sema/SemaDecl.cpp#L17616), causing `ParseEnumSpecifier` to [skip the body and return early](https://github.com/llvm/llvm-project/blob/cd708029e0b2869e80abe31ddb175f7c35361f90/clang/lib/Parse/ParseDecl.cpp#L5553-L5561) before it would otherwise [call `ParseEnumBody`](https://github.com/llvm/llvm-project/blob/cd708029e0b2869e80abe31ddb175f7c35361f90/clang/lib/Parse/ParseDecl.cpp#L5603) and create the `EnumDeclConstant` AST node (that we need to clean up in the C/Objective-C case).

2. Another difference, which doesn't seem to contribute directly to the divergence in this specific case, is that in C++, `ASTReader` pre-loads a set of interesting identifiers, see 33e0f7ef94124a53404555879666cef0e370f8c2.

> I wanted to try defining the enums in different scopes (e.g. in a struct) and see how it works. Consequences for scopes aren't clear, so that's worth taking another look.

I added one more test case for this. I got the same stack trace, and the crash seems to be fixed after the PR as well. However, I couldn’t reproduce the stack trace with `WriteCUDAPragmas`, so I might be missing something. See the scoped example [on CE](https://godbolt.org/#z:OYLghAFBqd5QCxAFwE4FN0BoCWIIDGAtgIYDW6AgqsAM4gDkAtACIDCAspQNICiA%2BgCEAqgEkAMi34AVAJoAFXgFIAzCxboARgFdgDLAQD2RAA44ANulTiSAO2DaSwdKIAmIAkoBMg74IPatMjGAPLayCbhAGIW6LYkROgo6EFYAGax9LDorjjBqG4gXumxhSollvGJILQIJBiuAHQIWKAM%2Bji0orYE5tquOYwADB20HCQ4tgDKhtqoBEkMAIyjISZxw/rmdsCF7QCUWBA5eYYF7uUZloUALBVxCUnoAB4JJpaNBCYmrYyj3b1%2BoMGCNcGMJtNZvNFqDOmsNiCtjs9gxDsdcvlCkt7oUAGz3KpJIiGVzaD7E1ykH5tf49PoDdyIsHjSYzOYLRgrMHw2ybLDbeyFTw%2BPxok6Y9x3K4udwAdgJjxAHBJlBiliWzV%2B7TBAPpwNh4NZUI5y1W615TIFu3cwt8PjFGLOhQArDj3AAOBXVZWuVWxLyamk6ulAxkGlmQ9mLLlw818q1Cvyio7ip3ufHSwoATi9SR9gjV6EDf2DgIZfM6EbZ0M5ZoRoITNqT9sOnU45HQ8lQhgAVugCMha1geyBsbZ0AB3QtYoaHcyMZ36C2gwyMaQYdAAam8Ki5g8RaKQRgGvEdqFw%2BGlhJqxqLCG%2B6UMtmQUwIJEso6lT%2BQwloneefIjsUVoeM2/gmHY6AAHKKrU9Q5Jqf6WAOOBPvgcSuGwhh9EQvIqHcGHiJM0HaEQmhWEUWAmIYtB5KhthYTheF3NRtHIPRRHjlBpHkaglFIf27FPlMyD1MgjGkcxWACShwmiagyCcSRZEUcUQRiRJuEgPh0nyYpxHcSpfFeIctC3owQSoNoA6bhwACeIlWTZSiyr4QyUJum5xKRtl2bwtikc626ub5ABq778CwwVRTE6DmK4qhuZQLksIlqL8guS58qucENJq%2B4jIeHgkugp6nOeeAQFeiqkJMnwPmk36vu%2BSRLF%2Bz6/v%2BgHafyOygSKPhUZBMHVC8bwfF8PwyUJvLHLYmHYZJo5YIRBk8RRLE0XRT6aby2KsdtthKYZvEgHc030SJGmLVp2IXXJYnHetfHnXpu3LepClPUZZ2meZDA7pMZZbt4Xj5oWzSg0o7nQ5QkzIJutW2BA%2BzBYIm4YMgcy2GFEWpSo6MpbD6Xzgwi5YMu%2BirmNpgTQ%2BBXpRAR4lWV%2BQXlVsTXqaj7Ps1H5tTzP5/vIAFMiO2IgbafhDVxNUkmSRYUlS0lxYJ9HofN73YqtXHPaO%2BIHTNWtUVtM3fadSz4vd0xvTde0q8hM1XV9a0/ZbunXUxH16ebFGW39UaMBSCu2SqaOw55weWKHvqFks4fuZ5nkIOgJADKg25eGDKpx5DWcR0nLzUQpm4AFQF0TieI/L0c%2Bn6lheAnHlJynadWJn2ex/6edeAXnlF2cCPl1XlfJbKqUw%2B5UdbvmTfJ6n6cd%2BDsQ9wXA8l8PY8T5QJOZRT2UgNPjRKyQPwM0Vx6lWe7PVdU3ONbzb78%2B1QtdWLZ29fY/V2uBw2KnXucWjWw1gtL22t5q%2B2MibNi9FjaGw4q7U6alVayRtp7Jad0UFOx9og1SHsFLG0%2BvpXWP0TLSX%2BoDEMAwO65QQggKG7ld5kyykyVcADYgahaOfI4zMTzX0qrfaM%2BgH4vifq1F%2BnURbdVdJLMCMtoL/xzt3IBWD1ZzVARglaEDcFQPgTtO2y09FHR0fxVRD0CEGMwY7S6OCSFIPweJSxDjIFFADjWBgAB6Dxm5eCmGQHZTc0piZzj3pTLAbClEN3ypsC%2BLN%2BGXk5oqe%2BTUxGfkFpI0WoIRz4lkQNX%2BstvQkgLCvFR1i0LqONjrZSp1XRGLgabBBdiKKumts7RxYCHZq3McQ6pzSHGENsb0vizo3Emi8THYp0cBjmlcLQTcT4Y713QAGBA8zbDmDso0WGlDgZL0ics1e7lxm6n6JMYAm5aFNFWSnDAm43yBBSJuKY6BSCbmCIjKwzg3kp03JQKY0hNy2BKnMtI3YiAXLqHlVZE48irOQD8hAOBXADBxk%2BR5oLjCLK2TDLwKggZ6hoZCuhDCd4hOYfvVhSoikQy4TEw4rd074EOGQEAzohgZQYDcfQRBRxDHZWE1c2p6DYm4XAWAMBEDFVMLECq%2BAjDSssHxXoOx%2BAkFoH%2BBS9FaD8DQNoWwZAKjICsPQCAmg%2BSaEmPUOyQ4LXxFQHZEImg%2BwDiHPKxIz4QjrOtUyAYOhgBsHfOYeg3MBikHsB%2BH1OAMAoQAG4pD5C8fs4QhG4GfHFeMOBNCoCtVhAYfI0A4B5dzONqBNA0XQBoMNwBzBnJiekbYdBQo4EnDyIcIi%2BbiPScLTJ%2Bhxaf1AFLQaNbNBZBRuEkwM1g1MGeHczcTA0jTzmfO6eTAqTztiKoFgR8T4mDnWkWgdlnwkGeEwJ8GyqYltQEi%2BNkBTJmNmhhd6NxXRVJOhRcodSDHPugYdFx5RWm2y9t%2BgDj0TH/sA0tYDgy318RUIcNAmBRxkvJgKxgm5njulxEwXENw7lWk3BANVGrJ1vKsvq1GEB1yYEzrufY%2Bhz70oXhRMdpMuVYB5Sod0jRcWylxLKJYspMPOndEMG4KhXSoaFSAdloqJVMxALMCI4R2byrMIqkaiwMNYZw3hnYm4lhZkaBqUE7bUkCxERk7qIqNxIuWhObNJhazIZYSuRgYQlMI0MGkdDmHsO4eVfYAjRGrAkd1eRgjWEFXtx3EsOjWAGMcrYzy3EGpeO4q8O6LwSx8IGedPiST%2Bh6AybpVgFlbKOUqBc1TEsCXSZeCq%2BEmrJWS20TQjcIAA%3D).

michael-jabbour-sonarsource commented:
@vsapsai Thank you for sharing the previous patch. This is really interesting.

> For the record, the previous work I've abandoned is https://reviews.llvm.org/D114833 Doesn't seem particularly relevant to this change to be honest.

As far as I can see, it actually seems to introduce the exact same `IdResolver.RemoveDecl(ECD)` call in `Sema::ActOnDuplicateDefinition` that I am introducing in this PR, and this seems to be sufficient to fix the crash here as well.

I tried to understand why that work was abandoned, in order to know if the reasoning applies to the changes in my PR.

- Regarding the "ambiguous use of internal linkage declaration" warning in the C++ test ([this comment](https://reviews.llvm.org/D114833#3165889)), I could see that the currently the warning is emitted before and after my PR on that test, so I see no risk here.
- About handling anonymous enums ([this comment](https://reviews.llvm.org/D114833#3190487)), do you think any further action needs to be taken? I have a test case to show how their AST dump looks like...

vgvassilev commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -2639,6 +2628,19 @@ void Sema::MergeTypedefNameDecl(Scope *S, TypedefNameDecl *New,
   notePreviousDefinition(Old, New->getLocation());
 }
 
+void Sema::CleanupMergedEnum(Scope *S, Decl *New) {
```
I feel like this is a property of merging. Can we find a common path when merging definitions to put that logic there? Is `MergeTypedefNameDecl` not called for the motivating example?

michael-jabbour-sonarsource replied:
> Is MergeTypedefNameDecl not called for the motivating example?

Unfortunately, it seems to me that `MergeTypedefNameDecl` is only called in the typedef to anonymous enum case (`MyEnum3` in the test case I am adding). For the rest of the enum cases, the closest I found was `Sema::ActOnDuplicateDefinition`, which is called when merging all tags (and this is where I am adding the new call to `CleanupMergedEnum`).

> Can we find a common path when merging definitions to put that logic there?

I could only see that merging enums for C and Obj-C crashes in this case (C++ works differently, see [here](https://github.com/llvm/llvm-project/pull/114240#issuecomment-2614544626)), and I found `Sema::ActOnDuplicateDefinition` during my investigation to be the function that handles merging tags.

I am not aware of a central place for merging definitions in general. Could you provide some hints on what to look for?

michael-jabbour-sonarsource commented:
Gentle ping :smile:

@vsapsai, @vgvassilev, I think I have responded to your inquiries and added a few related questions. Could you have a look?

vgvassilev reviewed (approved):
LGTM!

vgvassilev commented:
Please rebase this PR.

michael-jabbour-sonarsource committed:
[clang] Fix ASTWriter crash after merging named enums
